### ISSUE TITLE
GitHub Enterprise authentication with self signed certificate

### ISSUE THREAD

**Issue by nvs-abhilash** [jupyterhub/oauthenticator#253]
I was trying to setup authentication on JupyterHub using GitHub Enterprise. I have provided my configuration file.

I run jupyterhub by: `sudo jupyterhub`. After I log in with GitHub Enterprise credentials, I see 500 Server Error on the client. And on the server, I see the following error:
```
tornado.curl_httpclient.CurlError: HTTP 599: server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none
```

I suspect, it is because my GitHub Enterprise is hosted with a self signed certificate. Is there a way to get past this? (I know I get get the certificates if I have a domain, but currently I don't).

I saw somthing like `validate_server_cert` in GitLab code. Should we have something similar in GitHub for GitHub Enterprise deployments with self signed certificates?

My conf:
```python
import os
os.environ['GITHUB_HOST'] = "<GITHUB-HOST IP>"
# os.environ['GITHUB_HTTP'] = "1" # Tried with and without this

from oauthenticator.github import GitHubOAuthenticator

c.JupyterHub.ip = '<jupyterhub-host>'
c.JupyterHub.port = <jupyterhub-port>

c.JupyterHub.authenticator_class = GitHubOAuthenticator
c.GitHubOAuthenticator.oauth_callback_url = 'https://<jupyterhub-ip>:<jupyterhub-port>/hub/oauth_callback'
c.GitHubOAuthenticator.client_id = '<client_id>'
c.GitHubOAuthenticator.client_secret = '<client_secret>'

c.LocalAuthenticator.create_system_users = True
c.Authenticator.whitelist = {'myuser'}
c.Authenticator.admin_users = {'myuser'}
c.JupyterHub.ssl_cert = 'certs/mycert.pem'
c.JupyterHub.ssl_key = 'certs/mykey.key'
```

**Comment by sampathkethineedi** [comment]
I have a self signed certificiate on gitlab instance and I'm facing a similar issue during sign-in. Do we add the certificate to the pod when pulling the docker image?

Please help. Thanks!

**Comment by sampathkethineedi** [comment]
I used a custom image for hub built from the default jhub-k8s image. Added the self signed certificate and updated the ca certs. Worked for me.

**Comment by dmpe** [comment]
Having a custom image too, is there any way how can I import cert into docker image and avoid the error. 

I want to use GitLab and in my image I already do:

```
COPY ca-bundle.crt /etc/ssl/certs/ca-certificates.crt
....
RUN (set -e; \
......
     update-ca-certificates; \
)
```

**Comment by sampathkethineedi** [comment]
Hi @dmpe 

You can add these commands to the existing jhub-k8s dockerfile, build and push it to a registry and add it to the hub.image in config.yaml file
``` 
COPY <YOUR>.crt /usr/local/share/ca-certificates/
RUN update-ca-certificates --fresh
```
**config.yaml:**

```
hub:
  image: <registry>/<account>/<image>:<tag>
```

**Comment by dmpe** [comment]
@sampathkethineedi I have just solved that also with this.. in docker compose file
```
    environment: 
      - OAUTH_TLS_VERIFY=0
```

for gitlab (and maybe others) it is quite poorly documented.....

**Comment by arichar6** [comment]
> @sampathkethineedi I have just solved that also with this.. in docker compose file
> 
> ```
>     environment: 
>       - OAUTH_TLS_VERIFY=0
> ```
> 
> for gitlab (and maybe others) it is quite poorly documented.....

This solution worked for me to get around this same error. In my case I'm not using a docker image, rather a standalone TLJH install. So I had to edit `/etc/systemd/system/jupyterhub.service` and add `Environment=OAUTH_TLS_VERIFY=0` in the `[Service]` section.

**Comment by nvs-abhilash** [comment]
Fixed in #384

**Comment by ableuler** [comment]
> Fixed in #384

For other people coming here I guess this should be #348.

**Comment by consideRatio** [comment]
I think it was meant to be closed by #354
